site/content/concepts/device-lifecycle-notifications.md

```diff
@@ -0,0 +1,93 @@
++++
+title = "Device lifecycle notifications (from 0.6 on)"
+weight = 415
++++
+
+This page describes how device lifecycle notifications are handled in the Eclipse Hono project.
+
+
+{{% warning %}}
+This concept is currently part of active development and discussion.
+The content here summarizes the current discussion and is being prototyped.
+Please join our mailing list or consult the IRC channel in case you have any questions or want to participate.
+{{% /warning %}}
+
+<!--more-->
+
+Business applications connected to Hono may want to invoke arbitrary functionality when a device connects or disconnects to Hono.
+This is especially necessary to implement Command and Control, but is not limited to this Use Case. Integrating digital twins, like
+[Eclipse Ditto] (https://projects.eclipse.org/proposals/eclipse-ditto), also benefit from knowing when a device just connected to Hono.
+
+For these scenarios Hono defines a lifecycle notification concept that covers different patterns that try to cover several classes
+of protocols and devices.
+
+# Covered patterns
+
+Due to Hono's support for arbitrary protocols and arbitrary *classes* of devices (from strictly constrained to powerful gateways),
+there is no single way of sending lifecycle notifications.
+
+To be distinguished at least are:
+
+1. Connections oriented protocols, like MQTT
+1. Connections less protocols, like HTTP
+1. Protocols that support the setting of headers (like HTTP), in contrast to others (like MQTT)
+1. Devices that stay connected and can be contacted upstream always (e.g. via MQTT)
+1. Devices that are strictly constrained and connect only for a very short time per day, being only able to receive a few
+bytes upstream then
+1. ... and probably more specialties of protocols and devices over the time
+
+Besides that devices that connect to Hono may often not be directly ready for receiving upstream data, but need to separate
+their availability from the connection itself.
+
+Considering these different patterns, the main principle for lifecycle notifications is the following:
+
+# Main principle
+
+**The standard responsibility to signal that a device is ready to receive upstream data is at the device itself.**
+
```

Comment:
This sounds a little demanding, don't you think?
Maybe we can describe this more like an offering, e.g.
>  A device MAY signal its ability to receive a command from an application.

And then elaborate on the different options and means how to do that.

Reply:
Sounds demanding, right - and it did its purpose :-) 

We can use the *offering* phrase, but we should clearly state that usually not the protocol adapters will take over the job to signal command readiness - it has to be done by the device itself. I will search for a phrasing *in between* these variants...

Reply:
I believe the same :+1: 

So far I have only thought about the events, and not even about the downstream messages.
But for the event signaling command readiness the same probably is necessary.

For now, I would only support
- devices can send the command readiness event
- devices can piggy-back telemetry/event data
Later both of that can be configured per tenant and protocol adapter to do that instead of the device (to support specific device setups that do not support it).